core.internal.elf.io: Fix segfaults after mmap errors

kinke opened the pull request:
`mmap()` errors lead to `TypedMMapRegion.data` being null, which previously wasn't handled and led to segfaults - while trying to
resolve DWARF file/line infos for an exception backtrace.

This backtrace code needs to be as resilient as possible, so give up on the first mmap error now.

thewilsonator commented:
Is there an issue number for this? 
should this go to stable?
can this be tested?

RazvanN7 commented on src/core/internal/elf/io.d:
```diff
@@ -120,19 +123,13 @@ template ElfIO(Elf_Ehdr, Elf_Shdr, ubyte ELFCLASS)
             else version (BigEndian) alias ELFDATA = ELFDATA2MSB;
             else static assert(0, "unsupported byte order");
 
-            if (fd == -1)
```
I don't see any point in moving this `if` at the start of the function. PR's should be minimal in their changes.

kinke commented:
> should this go to stable?

Yeah, rebased.

> Is there an issue number for this?

We've encountered non-reliably-reproducible segfaults at work in this code (by inspecting core dumps).
The SIGBUS issue for empty (and presumably very small) files is one aspect of https://issues.dlang.org/show_bug.cgi?id=21656.

> can this be tested?

I've added a testcase for SIGBUS with empty file. I expected mmap to fail for a readonly empty file, but nope, it succeeds and later leads to SIGBUS when accessing the memory (while verifying the ELF header).
